[FLINK-9574] [doc]  Rework documentation for custom state serializers and state evolution (#7124)
apache/flink: docs/dev/stream/state/custom_serialization.md

under the License.
-->

* ToC
{:toc}

This page is targeted as a guideline for users who require the use of custom serialization for their state, covering how

> to provide a custom state serializer and guidelines and best practices to implementing serializers that allows

Reviewer: ```suggestion
how to provide a custom state serializer as well as guidelines and best practices for implementing serializers that allow
```

</div>
</div>

## State serializers and schema evolution

This section explains the user-facing abstractions related to state serialization and schema evolution, and necessary
internal details about how Flink interacts with these abstractions.

> When restoring from savepoints, Flink allows changing the serializers used to read and write prior registered state,

Reviewer: ```suggestion
When restoring from savepoints, Flink allows changing the serializers used to read and write previously registered state
```

Reviewer: I suggest changing all instances of "prior registered state" to "previously registered state", and "prior serializer" to "previous serializer". I find this wording more natural.

so that users are not locked in to any specific serialization schema. When state is restored, a new serializer will be
registered for the state (i.e., the serializer that comes with the `StateDescriptor` used to access the state in the
restored job). This new serializer may have a different schema than that of the prior serializer. Therefore, when
implementing state serializers, besides the basic logic of reading / writing data, another important thing to keep in
mind is how the serialization schema can be changed in the future.

When speaking of *schema*, in this context the term is interchangeable between referring to the *data model* of a state
type and the *serialized binary format* of a state type. The schema, generally speaking, can change for a few cases:

 1. Data schema of the state type has evolved, i.e. adding or removing a field from a POJO that is used as state.

>  2. Following the above, generally speaking, the serialization format of the serializer is intended to be upgraded.

Reviewer: Following the above, generally speaking, the serialization format of the serializer is intended to be upgraded.

I don't understand this sentence. Perhaps you mean to say that "Generally speaking, after a change to the data schema, the serialization format of the serializer will need to be upgraded."

### The `TypeSerializerSnapshot` abstraction

<div data-lang="java" markdown="1">
{% highlight java %}
public interface TypeSerializerSnapshot<T> {
    int getCurrentVersion();
    void writeSnapshot(DataOuputView out) throws IOException;
    void readSnapshot(int readVersion, DataInputView in, ClassLoader userCodeClassLoader) throws IOException;
    TypeSerializerSchemaCompatibility<T> resolveSchemaCompatibility(TypeSerializer<T> newSerializer);
    TypeSerializer<T> restoreSerializer();
}
{% endhighlight %}
</div>

<div data-lang="java" markdown="1">
{% highlight java %}
public abstract class TypeSerializer<T> {    
    
    // ...
    
    public abstract TypeSerializerSnapshot<T> snapshotConfiguration();
}
{% endhighlight %}
</div>

A serializer's `TypeSerializerSnapshot` is a piece of information that serves as the single source of truth about
the state serializer's configuration and write schema. The logic about what should be written and read at restore time
as the serializer snapshot is defined in the `writeSnapshot` and `readSnapshot` methods.

Note that the snapshot's own write schema may also need to change over time (e.g. when you wish to add more information

> about the serializer to the snapshot). To facilitate this, snapshot's are versioned, with the current version

Reviewer: ```suggestion
about the serializer to the snapshot). To facilitate this, snapshots are versioned, with the current version
```

> number defined in the `getCurrentVersion` method. On restore when the serializer snapshot is read from savepoints,

Reviewer: ```suggestion
number defined in the `getCurrentVersion` method. On restore, when the serializer snapshot is read from savepoints,
```

> the version of the schema that the snapshot was written in will be provided to the `readSnapshot` method so that

Reviewer: ```suggestion
the version of the schema in which the snapshot was written will be provided to the `readSnapshot` method so that
```

Another important role of a serializer's `TypeSerializerSnapshot` is that it serves as a factory to restore
the prior serializer. More specifically, the `TypeSerializerSnapshot` should implement the `restoreSerializer` method
to instantiate a serializer instance that recognizes the prior serializer's schema and configuration, and can therefore
safely read data written by the prior serializer.

### How Flink interacts with the `TypeSerializer` and `TypeSerializerSnapshot` abstractions

To wrap up, this section concludes how Flink, or more specifically the state backends, interact with the
abstractions. The interaction is slightly different depending on the state backend, but this is orthogonal
to the implementation of state serializers and their serializer snapshots.

#### Off-heap state backends (e.g. `RocksDBStateBackend`)

 1. **Register new state with a state serializer that has schema _A_**
  - the registered `TypeSerializer` for the state is used to read / write state on every state access.
  - State is written in schema *A*.
 2. **Take a savepoint**
  - The serializer snapshot is extracted via the `TypeSerializer#snapshotConfiguration` method.

>   - The serializer snapshot is written to the savepoint, as well as the already-deserialized state bytes (with schema *A*).

Reviewer 2: did you mean: already-serialized?

 3. **Restored execution re-accesses restored state bytes with new state serializer that has schema _B_**
  - The previous state serializer's snapshot is restored.
  - State bytes are not deserialized on restore, only loaded back to the state backends (therefore, still in schema *A*).
  - Upon receiving the new serializer, it is provided to the restored prior serializer's snapshot via the
  `TypeSerializer#resolveSchemaCompatibility` to check for schema compatibility.
 4. **Migrate state bytes in backend from schema _A_ to schema _B_**

>   - If the compatibility resolution reflects that the schema has changed and migration is possible, schema migration is 

Reviewer 3: I would make it more explicit that the whole state backend is migrated at once rather than in a lazy fashion.

  performed. The prior state serializer which recognizes schema _A_ will be obtained from the serializer snapshot, via
   `TypeSerializerSnapshot#restoreSerializer()`, and is used to deserialize state bytes to objects, which in turn

>    is re-written again with the new serializer, which recognizes schema _B_ to complete the migration.

Reviewer: ```suggestion
   are re-written again with the new serializer, which recognizes schema _B_ to complete the migration.
```

  - Upon receiving the new serializer, it is provided to the restored prior serializer's snapshot via the
  `TypeSerializer#resolveSchemaCompatibility` to check for schema compatibility.
  - If the compatibility check signals that migration is required, nothing happens in this case since for
   heap backends, all state is already deserialized into objects.
  - If the resolution signals incompatibility, then the state access fails with an exception.
 5. **Take another savepoint, serializing all state with schema _B_**
  - Same as step 2., but now state bytes are all in schema _B_.

## Implementation notes and best practices

#### 1. Flink restores serializer snapshots by instantiating them with their classname

A serializer's snapshot, being the single source of truth for how a registered state was serialized, serves as an
entrypoint to reading state in savepoints. In order to be able to restore and access previous state, the prior state
serializer's snapshot must be able to be restored.

Flink restores serializer snapshots by first instantiating the `TypeSerializerSnapshot` with its classname (written

> along with the snapshot bytes). Therefore, to avoid being subject to unintended classname changes,

Reviewer 3: I would list those requirements in a bulleted list. I find lists easier to follow, plus one can treat them as check lists.

Moreover, the class requires having a public, nullary constructor for instantiation.

#### 2. Avoid sharing the same `TypeSerializerSnapshot` class across different serializers

Since schema compatibility checks goes through the serializer snapshots, having multiple serializers returning
the same `TypeSerializerSnapshot` class as their snapshot would complicate the implementation for the
`TypeSerializerSnapshot#resolveSchemaCompatibility` and `TypeSerializerSnapshot#restoreSerializer()` method.

This would also be a bad separation of concerns; a single serializer's serialization schema,
configuration, as well as how to restore it, should be consolidated in its own dedicated `TypeSerializerSnapshot` class.

#### 3. Use the `CompositeSerializerSnapshot` utility for serializers that contain nested serializers

There may be cases where a `TypeSerializer` relies on other nested `TypeSerializer`s; take for example Flink's
`TupleSerializer`, where it is configured with nested `TypeSerializer`s for the tuple fields. In this case,
the snapshot of the most outer serializer should also contain snapshots of the nested serializers.

> The `ComnpositeSerializerSnapshot` can be used specifically for this scenario. It wraps the logic of resolving

Reviewer 3: ```suggestion
The `CompositeSerializerSnapshot` can be used specifically for this scenario. It wraps the logic of resolving
```
